crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs

```diff
@@ -315,22 +321,24 @@ impl GroupSessionManager {
             "Created a to-device request carrying a room_key"
         );
 
-        Ok((txn_id, request, share_infos, changed_sessions))
+        Ok((txn_id, request, share_infos, changed_sessions, withheld_devices))
     }
 
     /// Given a list of user and an outbound session, return the list of users
     /// and their devices that this session should be shared with.
     ///
-    /// Returns a boolean indicating whether the session needs to be rotated and
-    /// the list of users/devices that should receive the session.
+    /// Returns information indicating whether the session needs to be rotated
+    /// and the list of users/devices that should receive or not the session
+    /// (with withheld reason).
     pub async fn collect_session_recipients(
         &self,
         users: impl Iterator<Item = &UserId>,
         settings: &EncryptionSettings,
         outbound: &OutboundGroupSession,
-    ) -> OlmResult<(bool, HashMap<OwnedUserId, Vec<Device>>)> {
-        let users: HashSet<&UserId> = users.collect();
-        let mut devices: HashMap<OwnedUserId, Vec<Device>> = HashMap::new();
+    ) -> OlmResult<CollectRecipientsResult> {
+        let users: BTreeSet<&UserId> = users.collect();
+        let mut devices: BTreeMap<OwnedUserId, Vec<Device>> = Default::default();
+        let mut withheld_devices: Vec<(Device, WithheldCode)> = Vec::new();
```

Comment:
A bit weird that one is initialized with `Type::new` and the other with `Default::default` when both are possible in both cases 😄